## Playstore Apps analysis & Visualization

## About the project:

***In this project, we will be working on a real-world dataset of the google play store, one of the most used applications for downloading android apps. This project aims on cleaning the dataset, analyze the given dataset, and mining informational quality insights. This project also involves visualizing the data to better and easily understand trends and different categories.***

### Tasks performed in this notebook:

***1.Raw Data Importing***

***2.Raw Data Cleaning (& analyzing)***

***3.Cleaned Data Exporting***

***We have performed these tasks for both of the datsets i.e. playstore_apps.csv and playstore_reviewss.csv***

In [1]:
# Installing Textblob to perform Sentiment Analysis
!pip install textblob

In [2]:
# Importing necessary libraries and packages
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from textblob import TextBlob
import warnings
warnings.filterwarnings('ignore')

In [113]:
reviews=pd.read_csv("playstore_reviews.csv")

In [3]:
# Readind the csv file as df
df = pd.read_csv('googleplaystore.csv')
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [4]:
# Finding the total number of Rows and Columns in Dataset
df.shape

(10841, 13)

In [5]:
# Looking at information of apps dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [6]:
# Getting a description of the Dataset
df.describe()

Rating
count  9367.000000
mean      4.193338
std       0.537431
min       1.000000
25%       4.000000
50%       4.300000
75%       4.500000
max      19.000000

In [7]:
# Looking for duplicate records

df.duplicated().sum()

483

In [8]:
df.isnull().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

In [9]:
# dropping all duplicates, but keeping the first one

df.drop_duplicates(keep="first", inplace=True)

In [10]:
# cross-checking duplicate records

df.duplicated().sum()

0

In [11]:
# looking for number of unique values in each columns

df.nunique()

App               9660
Category            34
Rating              40
Reviews           6002
Size               462
Installs            22
Type                 3
Price               93
Content Rating       6
Genres             120
Last Updated      1378
Current Ver       2832
Android Ver         33
dtype: int64

### App column

***App column has only 9660 unique values out of 10358 records. So, we need to drop these duplicates. That means we have 698 duplicates.***

In [12]:
# Transforming the App column to a proper format

df['App'] = df['App'].str.title()

In [13]:
# Lets check for the number of unique values present in the app column after giving a proper formatting to App names.

len(df['App'].unique())

9639

In [14]:
# Lets check for the number of duplicates present in the app column after giving a proper formatting to App names.

df['App'].duplicated().sum()

719

In [15]:
# Just overviewing the duplicates

df[df['App'].duplicated()]

App       Category  Rating  \
285                Quick Pdf Scanner + Ocr Free       BUSINESS     4.2   
293      Officesuite : Free Office + Pdf Editor       BUSINESS     4.3   
294                                       Slack       BUSINESS     4.4   
382    Messenger – Text And Video Chat For Free  COMMUNICATION     4.0   
383               Imo Free Video Calls And Chat  COMMUNICATION     4.3   
...                                         ...            ...     ...   
10647                         Motorola Fm Radio  VIDEO_PLAYERS     3.9   
10715                        Farmersonly Dating         DATING     3.0   
10720        Firefox Focus: The Privacy Browser  COMMUNICATION     4.4   
10730                               Fp Notebook        MEDICAL     4.5   
10753            Slickdeals: Coupons & Shopping       SHOPPING     4.5   

        Reviews                Size        Installs  Type Price  \
285       80804  Varies with device      5,000,000+  Free     0   
293     1002859                 35M    100,000,000+  Free     0   
294       51510  Varies with device      5,000,000+  Free     0   
382    56646578  Varies with device  1,000,000,000+  Free     0   
383     4785988                 11M    500,000,000+  Free     0   
...         ...                 ...             ...   ...   ...   
10647     54815  Varies with device    100,000,000+  Free     0   
10715      1145                1.4M        100,000+  Free     0   
10720     36981                4.0M      1,000,000+  Free     0   
10730       410                 60M         50,000+  Free     0   
10753     33599                 12M      1,000,000+  Free     0   

      Content Rating                   Genres       Last Updated  \
285         Everyone                 Business  February 26, 2018   
293         Everyone                 Business     August 2, 2018   
294         Everyone                 Business     August 2, 2018   
382         Everyone            Communication     August 1, 2018   
383         Everyone            Communication       June 8, 2018   
...              ...                      ...                ...   
10647       Everyone  Video Players & Editors        May 2, 2018   
10715     Mature 17+                   Dating  February 25, 2016   
10720       Everyone            Communication       July 6, 2018   
10730       Everyone                  Medical     March 24, 2018   
10753       Everyone                 Shopping      July 30, 2018   

              Current Ver         Android Ver  
285    Varies with device        4.0.3 and up  
293             9.7.14188          4.1 and up  
294    Varies with device  Varies with device  
382    Varies with device  Varies with device  
383      9.8.000000010501          4.0 and up  
...                   ...                 ...  
10647  Varies with device  Varies with device  
10715                 2.2          4.0 and up  
10720                 5.2          5.0 and up  
10730           2.1.0.372          4.4 and up  
10753                 3.9          4.4 and up  

[719 rows x 13 columns]

In [16]:
# Looking for number of values for each unique value in the app column

df['App'].value_counts()

Roblox                                           9
8 Ball Pool                                      7
Bubble Shooter                                   6
Helix Jump                                       6
Zombie Catchers                                  6
                                                ..
Theme Android P For Lg G7 & V35                  1
Pistachio Launcher For Android P 9.0             1
Theme Android P Design For Lg V30                1
Pacify Exceed (Android P) - Theme For Xperia™    1
Ihoroscope - 2018 Daily Horoscope & Astrology    1
Name: App, Length: 9639, dtype: int64

In [17]:
# Selecting the App repeating most frequently for analysis purpose

df[df['App'] == "Roblox"]

App Category  Rating  Reviews Size      Installs  Type Price  \
1653  Roblox     GAME     4.5  4447388  67M  100,000,000+  Free     0   
1701  Roblox     GAME     4.5  4447346  67M  100,000,000+  Free     0   
1748  Roblox     GAME     4.5  4448791  67M  100,000,000+  Free     0   
1841  Roblox     GAME     4.5  4449882  67M  100,000,000+  Free     0   
1870  Roblox     GAME     4.5  4449910  67M  100,000,000+  Free     0   
2016  Roblox   FAMILY     4.5  4449910  67M  100,000,000+  Free     0   
2088  Roblox   FAMILY     4.5  4450855  67M  100,000,000+  Free     0   
2206  Roblox   FAMILY     4.5  4450890  67M  100,000,000+  Free     0   
4527  Roblox   FAMILY     4.5  4443407  67M  100,000,000+  Free     0   

     Content Rating                        Genres   Last Updated  \
1653   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
1701   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
1748   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
1841   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
1870   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
2016   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
2088   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
2206   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
4527   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   

       Current Ver Android Ver  
1653  2.347.225742  4.1 and up  
1701  2.347.225742  4.1 and up  
1748  2.347.225742  4.1 and up  
1841  2.347.225742  4.1 and up  
1870  2.347.225742  4.1 and up  
2016  2.347.225742  4.1 and up  
2088  2.347.225742  4.1 and up  
2206  2.347.225742  4.1 and up  
4527  2.347.225742  4.1 and up

## Observation:

***Here we can see that records having 'Roblox' as App name differ only in terms of Category and Number of Reviews. So, we are going to select only that record which has the highest number of reviews. The problem occurs because of the scrapping technique. Scrapping has been done at different point of time because of which the values of different columns for the same app name are different.***

In [18]:
# Sorting records with respect to the parameters provided in the list below...
# ...so that we can keep the first record on the basis of applied filters

df.sort_values(by = ['App', 'Reviews', 'Installs', 'Last Updated' ], ascending = False, inplace = True)

In [19]:
# Dropping the duplicate App names without dropping the first record

df.drop_duplicates('App', keep = 'first', inplace = True)

In [20]:
# Cross checking whether the duplicate app names have been removed or not.

df['App'].duplicated().sum()

0

## Category Column

In [21]:
# inspecting category column

df['Category'].unique()

array(['ENTERTAINMENT', 'TOOLS', 'SOCIAL', 'LIFESTYLE', 'VIDEO_PLAYERS',
       'COMICS', 'FAMILY', 'BOOKS_AND_REFERENCE', 'COMMUNICATION',
       'MAPS_AND_NAVIGATION', 'HOUSE_AND_HOME', 'SHOPPING', 'GAME',
       'ART_AND_DESIGN', 'NEWS_AND_MAGAZINES', 'AUTO_AND_VEHICLES',
       'FINANCE', 'FOOD_AND_DRINK', 'PERSONALIZATION', 'SPORTS', 'DATING',
       'BUSINESS', 'PARENTING', 'HEALTH_AND_FITNESS', 'MEDICAL',
       'TRAVEL_AND_LOCAL', 'PHOTOGRAPHY', 'LIBRARIES_AND_DEMO', 'WEATHER',
       'PRODUCTIVITY', 'EVENTS', 'BEAUTY', 'EDUCATION', '1.9'],
      dtype=object)

## Observation:
***Looking at the above output, we can see that '1.9' is not of categorical datatype. So let's inspect into it.***

In [22]:
# Looking at the record with category as '1.9'

df[df['Category']=='1.9']

App Category  Rating Reviews  \
10472  Life Made Wi-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

## Observation:
***We can see that the data starting from the category column till the Android Version has data from the immediate column to its right. But even if we try to format the data by shifting the values, most of the columns would have NaN values. Hence dropping the entire row would be a better option.***

In [23]:
# dropping the record

df.drop(10472,inplace=True)

In [24]:
# Cross checking for unique values in Category column

df['Category'].unique()

array(['ENTERTAINMENT', 'TOOLS', 'SOCIAL', 'LIFESTYLE', 'VIDEO_PLAYERS',
       'COMICS', 'FAMILY', 'BOOKS_AND_REFERENCE', 'COMMUNICATION',
       'MAPS_AND_NAVIGATION', 'HOUSE_AND_HOME', 'SHOPPING', 'GAME',
       'ART_AND_DESIGN', 'NEWS_AND_MAGAZINES', 'AUTO_AND_VEHICLES',
       'FINANCE', 'FOOD_AND_DRINK', 'PERSONALIZATION', 'SPORTS', 'DATING',
       'BUSINESS', 'PARENTING', 'HEALTH_AND_FITNESS', 'MEDICAL',
       'TRAVEL_AND_LOCAL', 'PHOTOGRAPHY', 'LIBRARIES_AND_DEMO', 'WEATHER',
       'PRODUCTIVITY', 'EVENTS', 'BEAUTY', 'EDUCATION'], dtype=object)

In [25]:
# Checking the number of null values present in Category Column

df['Category'].isnull().sum()

0

### Rating column

In [26]:
# Checking for the number of null values in the Rating column

df['Rating'].isnull().sum()

1458

In [27]:
# Capturing the columns with null values in a list

features_with_nan=[i for i in df.columns if df[i].isnull().sum()>=1]
features_with_nan

['Rating', 'Type', 'Current Ver', 'Android Ver']

In [28]:
# Checking the percentage of null values present in each columns with null values

for feature in features_with_nan:
    print(feature," has", np.round(df[feature].isnull().mean()*100, 2), '% missing values')

Rating  has 15.13 % missing values
Type  has 0.01 % missing values
Current Ver  has 0.08 % missing values
Android Ver  has 0.02 % missing values


***Here we can see that Rating has 15.13 % of its total data as null. 15.13% is a significant number and such a huge number of missing values can influence our analysis greatly as we keep progressing. Therefore, either we should impute the values or we should drop them.***

***But we can not directly conclude neither to drop the null values nor to impute ithem.***

***In real world scenarios, even big companies struggle to gather enough amount of data for analysis purposes. In our case, we may lose some crucial informations from our dataset. So what should we do? Should we drop or not?***

***In real world screnarios, the distributions of the datset are generally skewed. But if we find too much skewness in our dataset, then our statistical models won’t work effectively.***

***In skewed data, the tail region (that can be seen in the distribution diagram) may act as an outlier for our statistical model, and we know that outliers adversely affect a model’s performance, especially regression-based models. While there are statistical models that are robust enough to handle outliers like tree-based models, we’ll be limited in what other models we can try. So what should we do? In this case, we’ll need to transform the skewed data so that it becomes a Gaussian (or normal) distribution. Removing outliers and normalizing our data will allow us to experiment with more statistical models.***

***Overall, its a trial and error process. It's important to realize that data normalization isn't always necessary. In fact, sometimes it makes sense to do the opposite and add redundancy to a database. We should try out both conditions and check if our statistical model is performing better in first condition or in second condition. And finally, we should opt the process with a better performance over the another.***

In [29]:
# Imputing Nan values with 0 

# (however dataset distribution is better if we repalce by mean) - cross_verified through sns.df
df['Rating'] = df['Rating'].fillna(0)

In [30]:
# Checking the number of unique values present in Rating column

df['Rating'].unique()

array([4.7, 4. , 4.6, 3.8, 0. , 4.4, 4.3, 4.1, 3.7, 4.2, 4.5, 4.8, 3.2,
       3.6, 3.9, 4.9, 2.9, 3.5, 5. , 3.1, 2.8, 1.9, 3.4, 3.3, 2.7, 1.8,
       2.4, 2.3, 2.5, 3. , 2. , 2.6, 1. , 2.2, 2.1, 1.4, 1.6, 1.5, 1.2,
       1.7])

### Reviews column

In [31]:
# Checking the datatype of Reviews Column

df['Reviews'].unique()

array(['11661', '19', '22098', ..., '2490', '2019', '40467'], dtype=object)

In [32]:
# Checking the datatype of Reviews Column

df['Reviews'].dtypes

dtype('O')

In [33]:
# Converting the datatype into int32
df['Reviews'] = df['Reviews'].astype("int32")

In [34]:
# Cross checking if the datatype is changed or not:

df['Reviews'].dtypes

dtype('int32')

In [35]:
# Having a glimpse on Reviews column

df['Reviews'].describe()

count    9.638000e+03
mean     2.164353e+05
std      1.832721e+06
min      0.000000e+00
25%      2.500000e+01
50%      9.690000e+02
75%      2.943075e+04
max      7.815831e+07
Name: Reviews, dtype: float64

### Size column

In [36]:
# Checking the unique values:

df['Size'].unique()

array(['4.0M', '3.2M', '18M', '26M', '59M', '34M', 'Varies with device',
       '21M', '30M', '15M', '17M', '9.3M', '8.5M', '22M', '41M', '7.3M',
       '14M', '6.8M', '4.5M', '40M', '8.0M', '2.7M', '16M', '12M', '9.8M',
       '32M', '24M', '3.8M', '1.7M', '36M', '549k', '13M', '11M', '6.7M',
       '19M', '118k', '72k', '2.5M', '5.6M', '52M', '25M', '5.9M', '37M',
       '4.6M', '8.2M', '20M', '62M', '23M', '94M', '66M', '50M', '45M',
       '75M', '96M', '38M', '3.5M', '78M', '544k', '84M', '7.2M', '7.1M',
       '6.1M', '33M', '61M', '9.0M', '3.0M', '72M', '3.6M', '5.7M', '67M',
       '98M', '47M', '7.7M', '27M', '7.5M', '5.3M', '69M', '85M', '2.4M',
       '83M', '86M', '44M', '5.1M', '6.3M', '4.7M', '1.2M', '6.0M', '99M',
       '3.7M', '404k', '8.7M', '95M', '46M', '6.4M', '64M', '4.4M',
       '8.6M', '2.0M', '1.9M', '77M', '28M', '4.9M', '51M', '55M', '3.1M',
       '49M', '81M', '76M', '42M', '2.6M', '87M', '1020k', '1.5M', '53M',
       '29M', '5.0M', '3.3M', '4.1M', '1.0M'

### Observation:
***1.We can see a lot of values ending with unwanted characters. So we need to recognise and remove them.***

***2.We can see that the size of the app is given in MBs and KBs in terms of M and K respectively.Hence we need to standardize the data by converting MB into KB***.

***3.There are many values as "Varies with device". So we need to replace them by zero in order to convert the whole Size column into numerical datatype (since the column doesn't already have 0 as a value).***

In [37]:
# Capturing the last character from each values into a list

l=[]
for i in df["Size"][:]:
    if i[-1] not in l:
        l.append(i[-1])
l

['M', 'e', 'k']

In [38]:
# Storing the indexes of the values ending with 'M' in store1.

store1 = df[df["Size"].str[-1]=="M"].index
store1

Int64Index([  882,  7559,  2575,  4362,  6334,  6420,   313,  5698, 10109,
             1488,
            ...
             8219,  7738,  8483,  1393, 10252,  5940,  4636,   324,  8532,
             8884],
           dtype='int64', length=8100)

In [39]:
# Removing the last character from the size values captured in stored1

for i in store1:
    df["Size"].loc[i] = df["Size"].loc[i][:-1]

In [40]:
# Cross checking if we have successfully removed 'M':

df['Size'].unique()

array(['4.0', '3.2', '18', '26', '59', '34', 'Varies with device', '21',
       '30', '15', '17', '9.3', '8.5', '22', '41', '7.3', '14', '6.8',
       '4.5', '40', '8.0', '2.7', '16', '12', '9.8', '32', '24', '3.8',
       '1.7', '36', '549k', '13', '11', '6.7', '19', '118k', '72k', '2.5',
       '5.6', '52', '25', '5.9', '37', '4.6', '8.2', '20', '62', '23',
       '94', '66', '50', '45', '75', '96', '38', '3.5', '78', '544k',
       '84', '7.2', '7.1', '6.1', '33', '61', '9.0', '3.0', '72', '3.6',
       '5.7', '67', '98', '47', '7.7', '27', '7.5', '5.3', '69', '85',
       '2.4', '83', '86', '44', '5.1', '6.3', '4.7', '1.2', '6.0', '99',
       '3.7', '404k', '8.7', '95', '46', '6.4', '64', '4.4', '8.6', '2.0',
       '1.9', '77', '28', '4.9', '51', '55', '3.1', '49', '81', '76',
       '42', '2.6', '87', '1020k', '1.5', '53', '29', '5.0', '3.3', '4.1',
       '1.0', '56', '887k', '1.3', '10.0', '41k', '1.8', '903k', '61k',
       '29k', '26k', '1.4', '5.8', '253k', '4.2', '9.9', '7

In [41]:
# Storing the indexes of the values ending with 'k' in store1.

store2 = df[df["Size"].str[-1] == "k"].index
store2

Int64Index([ 5832,  4761, 10455,  3346, 10471, 10798,  6602,  1542,  8946,
             5451,
            ...
            10041, 10051,  4977,  4973,  4953,  4970,  4871,  6671,  4897,
             4541],
           dtype='int64', length=314)

In [42]:
# Removing the last character from the size values captured in stored2...
# ...and simultaneously converting values from KB to MB


for i in store2:
    df["Size"].loc[i] = round(float(df["Size"].loc[i][:-1])/1024,2)

In [43]:
# Cross checking if we have successfully removed 'k':


df['Size'].unique()

array(['4.0', '3.2', '18', '26', '59', '34', 'Varies with device', '21',
       '30', '15', '17', '9.3', '8.5', '22', '41', '7.3', '14', '6.8',
       '4.5', '40', '8.0', '2.7', '16', '12', '9.8', '32', '24', '3.8',
       '1.7', '36', 0.54, '13', '11', '6.7', '19', 0.12, 0.07, '2.5',
       '5.6', '52', '25', '5.9', '37', '4.6', '8.2', '20', '62', '23',
       '94', '66', '50', '45', '75', '96', '38', '3.5', '78', 0.53, '84',
       '7.2', '7.1', '6.1', '33', '61', '9.0', '3.0', '72', '3.6', '5.7',
       '67', '98', '47', '7.7', '27', '7.5', '5.3', '69', '85', '2.4',
       '83', '86', '44', '5.1', '6.3', '4.7', '1.2', '6.0', '99', '3.7',
       0.39, '8.7', '95', '46', '6.4', '64', '4.4', '8.6', '2.0', '1.9',
       '77', '28', '4.9', '51', '55', '3.1', '49', '81', '76', '42',
       '2.6', '87', 1.0, '1.5', '53', '29', '5.0', '3.3', '4.1', '1.0',
       '56', 0.87, '1.3', '10.0', 0.04, '1.8', 0.88, 0.06, 0.03, '1.4',
       '5.8', 0.25, '4.2', '9.9', '73', '6.6', '57', '7.8', '9.4'

In [44]:
# Capturing string length for 'Varies with device' in a variable:

str_length = len("Varies with device")
str_length

18

In [45]:
# Storing the indexes of the values having "Varies with device" in the variable store3

store3 = df[df["Size"].str[-str_length:] == "Varies with device"].index
store3

Int64Index([9222, 3824, 3960, 2758, 6245, 7077, 1641,   89, 1370, 1914,
            ...
            1437, 4863, 7006, 4568, 1267, 4875, 3151, 3448, 7330, 7338],
           dtype='int64', length=1224)

In [46]:
# Replacing "Varies with device" with 0

for i in store3:
    df["Size"].loc[i] = 0

In [47]:
# Cross checking if we have successfully imputed the values or not:

df['Size'].unique()

array(['4.0', '3.2', '18', '26', '59', '34', 0, '21', '30', '15', '17',
       '9.3', '8.5', '22', '41', '7.3', '14', '6.8', '4.5', '40', '8.0',
       '2.7', '16', '12', '9.8', '32', '24', '3.8', '1.7', '36', 0.54,
       '13', '11', '6.7', '19', 0.12, 0.07, '2.5', '5.6', '52', '25',
       '5.9', '37', '4.6', '8.2', '20', '62', '23', '94', '66', '50',
       '45', '75', '96', '38', '3.5', '78', 0.53, '84', '7.2', '7.1',
       '6.1', '33', '61', '9.0', '3.0', '72', '3.6', '5.7', '67', '98',
       '47', '7.7', '27', '7.5', '5.3', '69', '85', '2.4', '83', '86',
       '44', '5.1', '6.3', '4.7', '1.2', '6.0', '99', '3.7', 0.39, '8.7',
       '95', '46', '6.4', '64', '4.4', '8.6', '2.0', '1.9', '77', '28',
       '4.9', '51', '55', '3.1', '49', '81', '76', '42', '2.6', '87', 1.0,
       '1.5', '53', '29', '5.0', '3.3', '4.1', '1.0', '56', 0.87, '1.3',
       '10.0', 0.04, '1.8', 0.88, 0.06, 0.03, '1.4', '5.8', 0.25, '4.2',
       '9.9', '73', '6.6', '57', '7.8', '9.4', '10', '6.5', 0.64

In [48]:
# Checking the datatype of the Size column:

df['Size'].dtypes

dtype('O')

In [49]:
# Converting the datatype of the Size column:

df['Size'] = df['Size'].astype("float32")

In [50]:
# Cross checking the datatype of the Size column:
df['Size'].dtypes

dtype('float32')

### Installs column

In [51]:
#looking for null values:

df['Installs'].isnull().sum()

0

In [52]:
# Looking at unique values:

df['Installs'].unique()

array(['1,000,000+', '10,000+', '100,000+', '5,000,000+', '5,000+',
       '100+', '50,000+', '50,000,000+', '1,000+', '10,000,000+', '500+',
       '1+', '500,000+', '100,000,000+', '10+', '1,000,000,000+', '50+',
       '5+', '500,000,000+', '0+', '0'], dtype=object)

In [53]:
df['Installs'] = df['Installs'].str.replace('+', '')


In [54]:
# Checking the datatype:

df['Installs'].dtypes

dtype('O')

In [55]:
# Converting float64 into int32:

df['Installs'] = df['Installs'].str.replace(',', '').astype(int)


In [56]:
df['Installs'].unique()

array([   1000000,      10000,     100000,    5000000,       5000,
              100,      50000,   50000000,       1000,   10000000,
              500,          1,     500000,  100000000,         10,
       1000000000,         50,          5,  500000000,          0])

### Type column

In [57]:
# looking for null values:

df['Type'].isnull().sum()

1

In [58]:
# Looking for that particular record with null value in Type column

df[df['Type'].isnull() == True]

App Category  Rating  Reviews  Size  Installs  \
9148  Command & Conquer: Rivals   FAMILY     0.0        0   0.0         0   

     Type Price Content Rating    Genres   Last Updated         Current Ver  \
9148  NaN     0   Everyone 10+  Strategy  June 28, 2018  Varies with device   

             Android Ver  
9148  Varies with device

### Observation:
***This particular record is not much useful as apart from NaN in Type column the other columns like Reviews, Size, Installs doesn't have appropriate values which could be useful for future Analysis.***

In [59]:
# Dropping the record

df = df[~df['Type'].isnull()]

In [60]:
# Cross - checking for null value

df['Type'].isnull().sum()

0

In [61]:
# checking unique values

df['Type'].unique()

array(['Free', 'Paid'], dtype=object)

In [62]:
# Looking at counts of sub-type of Type column

df['Type'].value_counts()

Free    8884
Paid     753
Name: Type, dtype: int64

### Price Column

In [63]:
# looking for null value
df['Price'].isnull().sum()

0

In [64]:
#looking for unique value

df['Price'].unique()

array(['0', '$399.99', '$1.49', '$2.99', '$1.99', '$0.99', '$12.99',
       '$2.50', '$19.99', '$2.56', '$9.99', '$4.49', '$1.04', '$3.99',
       '$2.90', '$1.00', '$5.99', '$2.49', '$79.99', '$8.99', '$4.99',
       '$16.99', '$1.97', '$3.49', '$17.99', '$6.99', '$6.49', '$1.20',
       '$4.85', '$2.95', '$4.59', '$10.99', '$7.49', '$4.84', '$1.76',
       '$29.99', '$7.99', '$5.49', '$10.00', '$4.60', '$3.02', '$14.99',
       '$39.99', '$1.70', '$15.99', '$4.29', '$24.99', '$9.00', '$8.49',
       '$1.59', '$1.61', '$89.99', '$74.99', '$15.46', '$1.26', '$400.00',
       '$299.99', '$18.99', '$37.99', '$379.99', '$25.99', '$3.88',
       '$13.99', '$2.00', '$394.99', '$3.61', '$11.99', '$200.00',
       '$4.77', '$28.99', '$46.99', '$3.28', '$154.99', '$1.96', '$3.95',
       '$4.80', '$109.99', '$19.40', '$14.00', '$1.75', '$2.59', '$19.90',
       '$3.90', '$1.50', '$5.00', '$1.29', '$3.04', '$2.60', '$33.99',
       '$3.08'], dtype=object)

In [65]:
df['Price'] = df['Price'].str.replace('$', '')

In [66]:
#cross check 
df['Price'].unique()

array(['0', '399.99', '1.49', '2.99', '1.99', '0.99', '12.99', '2.50',
       '19.99', '2.56', '9.99', '4.49', '1.04', '3.99', '2.90', '1.00',
       '5.99', '2.49', '79.99', '8.99', '4.99', '16.99', '1.97', '3.49',
       '17.99', '6.99', '6.49', '1.20', '4.85', '2.95', '4.59', '10.99',
       '7.49', '4.84', '1.76', '29.99', '7.99', '5.49', '10.00', '4.60',
       '3.02', '14.99', '39.99', '1.70', '15.99', '4.29', '24.99', '9.00',
       '8.49', '1.59', '1.61', '89.99', '74.99', '15.46', '1.26',
       '400.00', '299.99', '18.99', '37.99', '379.99', '25.99', '3.88',
       '13.99', '2.00', '394.99', '3.61', '11.99', '200.00', '4.77',
       '28.99', '46.99', '3.28', '154.99', '1.96', '3.95', '4.80',
       '109.99', '19.40', '14.00', '1.75', '2.59', '19.90', '3.90',
       '1.50', '5.00', '1.29', '3.04', '2.60', '33.99', '3.08'],
      dtype=object)

In [67]:
# Looking at the description of the Price column

df['Price'].describe()

count     9637
unique      90
top          0
freq      8884
Name: Price, dtype: object

In [68]:
# Checking the datatype

df['Price'].dtypes

dtype('O')

In [69]:
# Converting the datatype into 'float32'

df['Price'] = df['Price'].astype("float32")

In [70]:
# Cross -checking the datatype

df['Price'].dtypes

dtype('float32')

### Content Rating Column

In [71]:
# looking for null values

df['Content Rating'].isnull().sum()

0

In [72]:
# Looking at unique values

df['Content Rating'].unique()

array(['Everyone', 'Mature 17+', 'Teen', 'Everyone 10+',
       'Adults only 18+', 'Unrated'], dtype=object)

In [73]:
# Looking at counts of sub-type of Content Rating column

df['Content Rating'].value_counts()

Everyone           7886
Teen               1034
Mature 17+          392
Everyone 10+        320
Adults only 18+       3
Unrated               2
Name: Content Rating, dtype: int64

### Genres column

In [74]:
# Looking for null values

df['Genres'].isnull().sum()

0

In [75]:
# Looking at unique values

df['Genres'].unique()

array(['Entertainment', 'Tools', 'Social', 'Lifestyle',
       'Video Players & Editors', 'Comics', 'Education',
       'Books & Reference', 'Communication', 'Maps & Navigation',
       'House & Home', 'Shopping', 'Simulation', 'Adventure',
       'Art & Design', 'Strategy', 'News & Magazines', 'Auto & Vehicles',
       'Action', 'Finance', 'Food & Drink', 'Arcade', 'Personalization',
       'Sports', 'Casino', 'Education;Education', 'Dating', 'Business',
       'Puzzle', 'Parenting;Education', 'Health & Fitness', 'Medical',
       'Educational;Education', 'Travel & Local', 'Photography',
       'Entertainment;Music & Video', 'Trivia', 'Libraries & Demo',
       'Casual', 'Weather', 'Productivity', 'Racing;Action & Adventure',
       'Events', 'Books & Reference;Creativity', 'Beauty',
       'Puzzle;Action & Adventure', 'Card', 'Board;Pretend Play', 'Word',
       'Puzzle;Brain Games', 'Role Playing', 'Board;Action & Adventure',
       'Parenting', 'Racing', 'Arcade;Action & Adventure'

### Observation:
***We can observe that some of the values have two genres separated with ';' (semicolon). So for further analysis, we will separate the them into two columns named as Genre_1 and Genre_2 and we will remove original Genres column.***

In [76]:
# Storing the values of genre before ';' in Genre 1

Genre_1 = df['Genres'].apply(lambda value : value.split(';')[0])


# Storing the values of genre after ';' in Genre 2
# Places where Genre_2 values are not present, substituting them with 'Not Applicable'

Genre_2 = df['Genres'].apply(lambda value : value.split(';')[1] if len(value.split(';')) > 1 else "Not Applicable")

In [77]:
# Inserting the Genre_1 and Genre_2 columns at column index 10 & 11 respectively

df.insert(10, 'Genre_1', Genre_1)
df.insert(11, 'Genre_2', Genre_2)

In [78]:
# Looking at the dimension of apps dataset after addition of new columns

df.shape

(9637, 15)

In [79]:
# Capturing the values of those indexes where Genre_1 and Genre_2 are same

apps_index = df[df['Genre_1'] == df['Genre_2']].index

In [80]:
# Replacing the values in Genre_2 column with 'Not Applicable' where we found that Genre_2 is the same value as in Genre_1

for i in apps_index:
    df['Genre_2'][i] = "Not Applicable"

In [81]:
# Dropping the original Genre column and save it separtely

Original_Genres = df.pop('Genres')

In [82]:
# Looking at the dimension of apps dataset

df.shape

(9637, 14)

In [83]:
# Looking at apps dataset

df

App       Category  Rating  \
882   🔥 Football Wallpapers 4K | Full Hd Backgrounds 😍  ENTERTAINMENT     4.7   
7559  📏 Smart Ruler ↔️ Cm/Inch Measuring For Homework!          TOOLS     4.0   
2575     💘 Whatslov: Smileys Of Love, Stickers And Gif         SOCIAL     4.6   
4362                                        💎 I'M Rich      LIFESTYLE     3.8   
6334                           뽕티비 - 개인방송, 인터넷방송, Bj방송  VIDEO_PLAYERS     0.0   
...                                                ...            ...     ...   
4636                                            /U/App  COMMUNICATION     4.7   
4541                                                .R          TOOLS     4.5   
324                         - Free Comics - Comic Apps         COMICS     3.5   
8532                     +Download 4 Instagram Twitter         SOCIAL     4.5   
8884              "I Dt" Fútbol. Todos Somos Técnicos.         SPORTS     0.0   

      Reviews  Size  Installs  Type      Price Content Rating  \
882     11661   4.0   1000000  Free    0.00000       Everyone   
7559       19   3.2     10000  Free    0.00000       Everyone   
2575    22098  18.0   1000000  Free    0.00000       Everyone   
4362      718  26.0     10000  Paid  399.98999       Everyone   
6334      414  59.0    100000  Free    0.00000     Mature 17+   
...       ...   ...       ...   ...        ...            ...   
4636      573  53.0     10000  Free    0.00000     Mature 17+   
4541      259   0.2     10000  Free    0.00000       Everyone   
324       115   9.1     10000  Free    0.00000     Mature 17+   
8532    40467  22.0   1000000  Free    0.00000       Everyone   
8884       27   3.6       500  Free    0.00000       Everyone   

                      Genre_1         Genre_2        Last Updated Current Ver  \
882             Entertainment  Not Applicable       July 14, 2018     1.1.3.2   
7559                    Tools  Not Applicable    October 21, 2017         1.0   
2575                   Social  Not Applicable       July 24, 2018       4.2.4   
4362                Lifestyle  Not Applicable      March 11, 2018       1.0.0   
6334  Video Players & Editors  Not Applicable       July 18, 2018       4.0.7   
...                       ...             ...                 ...         ...   
4636            Communication  Not Applicable        July 3, 2018       4.2.4   
4541                    Tools  Not Applicable  September 16, 2014      1.1.06   
324                    Comics  Not Applicable       July 13, 2018      5.0.12   
8532                   Social  Not Applicable      August 2, 2018        5.03   
8884                   Sports  Not Applicable     October 7, 2017        0.22   

       Android Ver  
882   4.0.3 and up  
7559    4.2 and up  
2575  4.0.3 and up  
4362    4.4 and up  
6334  4.0.3 and up  
...            ...  
4636    4.1 and up  
4541    1.5 and up  
324     5.0 and up  
8532    4.1 and up  
8884    4.1 and up  

[9637 rows x 14 columns]

In [84]:
# Checking information of apps dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9637 entries, 882 to 8884
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             9637 non-null   object 
 1   Category        9637 non-null   object 
 2   Rating          9637 non-null   float64
 3   Reviews         9637 non-null   int32  
 4   Size            9637 non-null   float32
 5   Installs        9637 non-null   int32  
 6   Type            9637 non-null   object 
 7   Price           9637 non-null   float32
 8   Content Rating  9637 non-null   object 
 9   Genre_1         9637 non-null   object 
 10  Genre_2         9637 non-null   object 
 11  Last Updated    9637 non-null   object 
 12  Current Ver     9629 non-null   object 
 13  Android Ver     9635 non-null   object 
dtypes: float32(2), float64(1), int32(2), object(9)
memory usage: 1.2+ MB


### Last Updated Column

In [85]:
# Looking for null values

df['Last Updated'].isnull().sum()

0

In [86]:
# Looking at datatype

df['Last Updated'].dtypes

dtype('O')

In [87]:
df['Last Updated'].unique()

array(['July 14, 2018', 'October 21, 2017', 'July 24, 2018', ...,
       'August 30, 2013', 'July 3, 2014', 'September 16, 2014'],
      dtype=object)

In [88]:
# changing datatype of Last Updated from object--->datetime format so that...
# ...we can sort duplicates with respect to Last Updated

df['Last Updated'] = df['Last Updated'].astype('datetime64[ns]')

In [89]:
df['Last Updated'].unique()

array(['2018-07-14T00:00:00.000000000', '2017-10-21T00:00:00.000000000',
       '2018-07-24T00:00:00.000000000', ...,
       '2013-08-30T00:00:00.000000000', '2014-07-03T00:00:00.000000000',
       '2014-09-16T00:00:00.000000000'], dtype='datetime64[ns]')

### Current Ver Column

In [90]:
# Looking for null values

df['Current Ver'].isnull().sum()

8

In [92]:
# Replacing the null values with "Unknown" as its is a categorical column.

df['Current Ver'].fillna("Unknown", inplace= True)

In [93]:
# Checking null values

df['Current Ver'].isnull().sum()

0

In [94]:
# Looking at unique values

df['Current Ver'].unique()

array(['1.1.3.2', '1.0', '4.2.4', ..., '1.1.06', '5.03', '0.22'],
      dtype=object)

In [95]:
# Looking at the count of sub-type of Current Ver column

df['Current Ver'].value_counts()

Varies with device    1051
1.0                    792
1.1                    258
1.2                    176
2.0                    148
                      ... 
17.4                     1
8.07.021                 1
1.34.1                   1
5.36                     1
0.22                     1
Name: Current Ver, Length: 2819, dtype: int64

### Android Ver Column

In [96]:
# Looking for null values

df['Android Ver'].isnull().sum()

2

In [97]:
# Replacing the null values with "Unknown" as its is a categorical column.

df['Android Ver'].fillna("Unknown", inplace= True)

In [98]:
# Checking for null values

df['Android Ver'].isnull().sum()

0

In [99]:
# Looking at unique values

df['Android Ver'].unique()

array(['4.0.3 and up', '4.2 and up', '4.4 and up', '3.0 and up',
       'Varies with device', '4.0 and up', '2.3 and up', '4.1 and up',
       '5.0 and up', 'Unknown', '7.0 and up', '4.3 and up', '3.2 and up',
       '2.1 and up', '2.3.3 and up', '7.0 - 7.1.1', '2.2 and up',
       '6.0 and up', '5.1 and up', '1.6 and up', '4.4W and up',
       '1.5 and up', '4.0.3 - 7.1.1', '2.0 and up', '2.0.1 and up',
       '8.0 and up', '1.0 and up', '3.1 and up', '5.0 - 8.0',
       '7.1 and up', '5.0 - 6.0', '4.1 - 7.1.1', '5.0 - 7.1.1',
       '2.2 - 7.1.1'], dtype=object)

In [100]:
# earlier (10841,13)

# df dataset new dimension
df.shape

(9637, 14)

### Sanctity Checks
***Few checks/conditions that need to be fulfilled:***

***1.Rating for the Apps should be less than 5.***

***2.Reviews should not be greater than Installs for any app.***

***3.Free apps should not be priced/charged.***

### 1) Rating for the apps should be less than 5

In [101]:
# Looking for records where rating for an app is greater than 5

df['Rating']>5

882     False
7559    False
2575    False
4362    False
6334    False
        ...  
4636    False
4541    False
324     False
8532    False
8884    False
Name: Rating, Length: 9637, dtype: bool

In [102]:
# Looking at number of records where an app has a rating > 5

df[df['Rating']>5].shape

(0, 14)

***No app has rating above 5.***

In [103]:
df['Rating'].describe()

count    9637.000000
mean        3.542638
std         1.574796
min         0.000000
25%         3.600000
50%         4.200000
75%         4.500000
max         5.000000
Name: Rating, dtype: float64

### Observation:
***All the Ratings are well within the defined limit i.e. less than 5.***

### 2) Reviews should not be greater than Installs for any app

In [104]:
# looking for records where Reviews for an app are greater than Installs

df[df['Reviews']>df['Installs']]

App         Category  Rating  Reviews  \
7402                Trovami Se Ci Riesci             GAME     5.0       11   
6508                          Sam.Bn Pro            TOOLS     0.0       11   
4550   Rmedus - 음성인식을 활용한 R 프로그래밍 실습 시스템           FAMILY     0.0        4   
5917                            Ra Ga Ba             GAME     5.0        2   
10697                            Mu.F.O.             GAME     5.0        2   
2454                 Kba-Ez Health Guide          MEDICAL     5.0        4   
9096                           Dz Puzzle           FAMILY     0.0       14   
8591                             Dn Blog           SOCIAL     5.0       20   
6700                    Brick Breaker Br             GAME     5.0        7   
5812             Ax Watch For Watchmaker  PERSONALIZATION     0.0        2   
4663       Alarmy (Sleep If U Can) - Pro        LIFESTYLE     4.8    10249   

        Size  Installs  Type  Price Content Rating          Genre_1  \
7402    6.10        10  Free   0.00       Everyone           Arcade   
6508    2.00        10  Paid   0.99       Everyone            Tools   
4550   64.00         1  Free   0.00       Everyone        Education   
5917   20.00         1  Paid   1.49       Everyone           Arcade   
10697  16.00         1  Paid   0.99       Everyone           Arcade   
2454   25.00         1  Free   0.00       Everyone          Medical   
9096   47.00        10  Paid   0.99       Everyone           Puzzle   
8591    4.20        10  Free   0.00           Teen           Social   
6700   19.00         5  Free   0.00       Everyone           Arcade   
5812    0.23         1  Paid   0.99       Everyone  Personalization   
4663    0.00     10000  Paid   2.49       Everyone        Lifestyle   

              Genre_2 Last Updated         Current Ver         Android Ver  
7402   Not Applicable   2017-03-11                 0.1          2.3 and up  
6508   Not Applicable   2015-03-27               1.0.0        4.0.3 and up  
4550   Not Applicable   2018-07-17               1.0.1          4.4 and up  
5917   Not Applicable   2017-02-08               1.0.4          2.3 and up  
10697  Not Applicable   2017-03-03                 1.0          2.3 and up  
2454   Not Applicable   2018-08-02              1.0.72        4.0.3 and up  
9096   Not Applicable   2017-04-22                 1.2          2.3 and up  
8591   Not Applicable   2018-07-23                 1.0          4.0 and up  
6700   Not Applicable   2018-07-23                 1.0          4.1 and up  
5812   Not Applicable   2017-08-18                 1.0          2.3 and up  
4663   Not Applicable   2018-07-30  Varies with device  Varies with device

In [105]:
# no. of records
df[df['Reviews']>df['Installs']].shape

(11, 14)

### Observation:
***There are 11 records where Reviews are greater than Installs which is not an ideal case, hence we'll be storing only those records where Reviews are less than or equal to Installs.***

In [106]:
# Storing only those records where Reviews for an app are less than or equal to (<=) Installs

df = df[df['Reviews']<= df['Installs']]

In [107]:
# Cross- checking for apps where Reviews are greater than Installs

df[df['Reviews']>df['Installs']].shape

(0, 14)

### 3) Free apps should not be priced

In [108]:
# Looking for records where Free apps that are Priced/Charged

df[(df['Type']=='Free') & (df['Price']>0)]

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genre_1, Genre_2, Last Updated, Current Ver, Android Ver]
Index: []

In [109]:
# Looking at number of records where ree apps that are Priced/Charged

df[(df['Type']=='Free') & (df['Price']>0)].shape

(0, 14)

### observation:
***No Free App has been charged, thus fulfilling our condition.***

***All the Sanctity checks has been well performed.***

### Final checks:

In [115]:
Apps=df

In [116]:
# Looking at information of apps dataset

Apps.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9626 entries, 882 to 8884
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   App             9626 non-null   object        
 1   Category        9626 non-null   object        
 2   Rating          9626 non-null   float64       
 3   Reviews         9626 non-null   int32         
 4   Size            9626 non-null   float32       
 5   Installs        9626 non-null   int32         
 6   Type            9626 non-null   object        
 7   Price           9626 non-null   float32       
 8   Content Rating  9626 non-null   object        
 9   Genre_1         9626 non-null   object        
 10  Genre_2         9626 non-null   object        
 11  Last Updated    9626 non-null   datetime64[ns]
 12  Current Ver     9626 non-null   object        
 13  Android Ver     9626 non-null   object        
dtypes: datetime64[ns](1), float32(2), float64(1), int32(2)

In [111]:
# Looking at the number of unique values for each column in apps dataset

df.nunique()

App               9626
Category            33
Rating              40
Reviews           5323
Size               274
Installs            20
Type                 2
Price               90
Content Rating       6
Genre_1             48
Genre_2              7
Last Updated      1377
Current Ver       2818
Android Ver         34
dtype: int64

### Reviews Dataset

In [114]:
# Looking at top 5 rows

reviews.head()

App                                  Translated_Review  \
0  10 Best Foods for You  I like eat delicious food. That's I'm cooking ...   
1  10 Best Foods for You    This help eating healthy exercise regular basis   
2  10 Best Foods for You                                                NaN   
3  10 Best Foods for You         Works great especially going grocery store   
4  10 Best Foods for You                                       Best idea us   

  Sentiment  Sentiment_Polarity  Sentiment_Subjectivity  
0  Positive                1.00                0.533333  
1  Positive                0.25                0.288462  
2       NaN                 NaN                     NaN  
3  Positive                0.40                0.875000  
4  Positive                1.00                0.300000

In [117]:
# Looking at information of reviews dataset

reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  object 
 1   Translated_Review       37427 non-null  object 
 2   Sentiment               37432 non-null  object 
 3   Sentiment_Polarity      37432 non-null  float64
 4   Sentiment_Subjectivity  37432 non-null  float64
dtypes: float64(2), object(3)
memory usage: 2.5+ MB


In [118]:
# Looking at dimension of the dataset

reviews.shape

(64295, 5)

In [119]:
# looking for duplicate vaules

reviews.duplicated().sum()

33616

### Not removing duplicates as a single app can have multiple reviews thus multiple entry of the same app.

In [120]:
# looking for null values

reviews.isnull().sum()

App                           0
Translated_Review         26868
Sentiment                 26863
Sentiment_Polarity        26863
Sentiment_Subjectivity    26863
dtype: int64

In [121]:
# Looking at number of unique values present in each column of reviews dataset 

reviews.nunique()

App                        1074
Translated_Review         27994
Sentiment                     3
Sentiment_Polarity         5410
Sentiment_Subjectivity     4474
dtype: int64

### App Column

In [122]:
# Looking at unique values

reviews['App'].unique()

array(['10 Best Foods for You', '104 找工作 - 找工作 找打工 找兼職 履歷健檢 履歷診療室',
       '11st', ..., 'Hotwire Hotel & Car Rental App',
       'Housing-Real Estate & Property', 'Houzz Interior Design Ideas'],
      dtype=object)

In [123]:
# Lookig at counts of sub-type of App column

reviews['App'].value_counts()

Angry Birds Classic                                  320
CBS Sports App - Scores, News, Stats & Watch Live    320
Bowmasters                                           320
8 Ball Pool                                          300
Helix Jump                                           300
                                                    ... 
Easy Healthy Recipes                                  31
Dresses Ideas & Fashions +3000                        31
Detector de Radares Gratis                            31
Easy Hair Style Design                                30
Drawing Clothes Fashion Ideas                         30
Name: App, Length: 1074, dtype: int64

### Translated_Review column

In [124]:
# Removing all the rows with null values as they would not help in analysis

reviews = reviews[~reviews['Translated_Review'].isnull()]

In [125]:
# Cross - checking for null values

reviews['Translated_Review'].isnull().sum()

0

In [126]:
# Lookig at unique values

reviews['Translated_Review'].unique()

array(['I like eat delicious food. That\'s I\'m cooking food myself, case "10 Best Foods" helps lot, also "Best Before (Shelf Life)"',
       'This help eating healthy exercise regular basis',
       'Works great especially going grocery store', ...,
       'Dumb app, I wanted post property rent give option. Website work. Waste time space phone.',
       'I property business got link SMS happy performance still guys need raise bar guys Cheers',
       'Useless app, I searched flats kondapur, Hyderabad . None number reachable I know flats unavailable would keep posts active'],
      dtype=object)

### Sentiment column

In [127]:
# Looking for null values

reviews['Sentiment'].isnull().sum()

0

In [128]:
# Looking at unique values

reviews['Sentiment'].unique()

array(['Positive', 'Neutral', 'Negative'], dtype=object)

In [129]:
# Looking at counts of sub-type of Sentiment column

reviews['Sentiment'].value_counts()

Positive    23998
Negative     8271
Neutral      5158
Name: Sentiment, dtype: int64

### Sentiment_Polarity column

In [130]:
# looking for null values

reviews['Sentiment_Polarity'].isnull().sum()

0

In [131]:
# Looking for unique values

reviews['Sentiment_Polarity'].unique()

array([ 1.        ,  0.25      ,  0.4       , ..., -0.52857143,
       -0.37777778,  0.17333333])

In [132]:
# Rounding the values to 3 decimal digits

reviews['Sentiment_Polarity'] = np.round(reviews['Sentiment_Polarity'], 3)

In [133]:
# Cross-checking the unique values

reviews['Sentiment_Polarity'].unique()

array([ 1.   ,  0.25 ,  0.4  , ...,  0.593, -0.988, -0.529])

In [134]:
# Looking at the description of Sentiment_Polarity column

reviews['Sentiment_Polarity'].describe()

count    37427.000000
mean         0.182166
std          0.351319
min         -1.000000
25%          0.000000
50%          0.150000
75%          0.400000
max          1.000000
Name: Sentiment_Polarity, dtype: float64

### Sentiment_Subjectivity column

In [135]:
# looking for null values

reviews['Sentiment_Subjectivity'].isnull().sum()

0

In [136]:
# Looking for unique values

reviews['Sentiment_Subjectivity'].unique()

array([0.53333333, 0.28846154, 0.875     , ..., 0.51145833, 0.7172619 ,
       0.2594697 ])

In [137]:
# Rounding the values to 3 decimal digits

reviews['Sentiment_Subjectivity'] = np.round(reviews['Sentiment_Subjectivity'], 3)

In [138]:
# Cross-checking the unique values 
reviews['Sentiment_Subjectivity'].unique()

array([0.533, 0.288, 0.875, 0.3  , 0.9  , 0.   , 0.6  , 0.1  , 0.867,
       0.511, 1.   , 0.667, 0.8  , 0.35 , 0.5  , 0.69 , 0.2  , 0.75 ,
       0.675, 0.356, 0.55 , 0.15 , 0.412, 0.525, 0.65 , 0.833, 0.467,
       0.367, 0.597, 0.45 , 0.735, 0.475, 0.4  , 0.683, 0.286, 0.622,
       0.78 , 0.717, 0.25 , 0.275, 0.633, 0.587, 0.143, 0.767, 0.739,
       0.302, 0.364, 0.05 , 0.455, 0.375, 0.478, 0.458, 0.536, 0.333,
       0.562, 0.324, 0.417, 0.775, 0.917, 0.542, 0.357, 0.792, 0.067,
       0.644, 0.7  , 0.517, 0.706, 0.648, 0.579, 0.479, 0.594, 0.378,
       0.705, 0.506, 0.575, 0.52 , 0.975, 0.572, 0.192, 0.697, 0.688,
       0.558, 0.85 , 0.513, 0.625, 0.401, 0.436, 0.42 , 0.581, 0.322,
       0.469, 0.425, 0.41 , 0.544, 0.368, 0.599, 0.607, 0.749, 0.555,
       0.564, 0.326, 0.418, 0.585, 0.653, 0.521, 0.651, 0.92 , 0.771,
       0.578, 0.737, 0.708, 0.637, 0.258, 0.583, 0.397, 0.718, 0.371,
       0.48 , 0.568, 0.249, 0.758, 0.61 , 0.483, 0.502, 0.733, 0.489,
       0.889, 0.54 ,

In [139]:
# Looking at description of Sentiment_Subjectivity column

reviews['Sentiment_Subjectivity'].describe()

count    37427.000000
mean         0.492770
std          0.259905
min          0.000000
25%          0.357000
50%          0.514000
75%          0.650000
max          1.000000
Name: Sentiment_Subjectivity, dtype: float64

### After cleaning both the files

In [140]:
# new dimension of apps dataset

Apps.shape

(9626, 14)

***older dimension (10841,13)***

In [141]:
# new dimension of reviews dataset

reviews.shape

(37427, 5)

***older dimension (64295,5)***

### converting new cleaned files to csv

In [142]:
# Converting apps to cleaned_playstore_apps.csv

Apps.to_csv('cleaned_playstore_apps.csv', header=True, index=False)

In [143]:
# Converting reviewss to cleaned_playstore_reviews.csv

reviews.to_csv('cleaned_playstore_reviews.csv', header=True, index=False)